tools/repack-kernel

```diff
@@ -0,0 +1,235 @@
+#!/bin/bash
+set -e
+
+if [ -n "$D" ]; then
+    set -x
+fi
+
+show_help() {
+    echo "usage: repack-kernel <command> <opts>"
+    echo ""
+    echo "handle extraction of the kernel snap to a workspace directory, and later"
+    echo "repacking it back to a snap."
+    echo ""
+    echo "   repack-kernel setup                        - setup system dependencies"
+    echo "   repack-kernel extract <snap-file> <target> - extract under <target> workspace tree"
+    echo "   repack-kernel prepare <target>             - prepare initramfs & kernel for repacking"
+    echo "   repack-kernel pack <target>                - pack the kernel"
+    echo "   repack-kernel cull-firmware <target>       - remove unnecessary firmware"
+    echo "   repack-kernel cull-modules <target>        - remove unnecessary modules"
+    echo ""
+}
+
+setup() {
+    if [ "$UID" != "0" ]; then
+        echo "run as root (only this command)"
+        exit 1
+    fi
+
+    # carries ubuntu-core-initframfs
+    add-apt-repository ppa:snappy-dev/image -y
+    apt update -y
+    
+    if ! apt install ubuntu-core-initramfs -y; then
+        echo "ubuntu-core-initramfs was not available in deb repository!"
+        exit 1
+    fi
+}
+
+get_kver() {
+    kerneldir="$1"
+    (
+        cd "$kerneldir"
+        #shellcheck disable=SC2010
+        ls "config"-* | grep -Po 'config-\K.*'
+    )
+}
+
+extract() {
+    local snap_file="$1"
+    local target="$2"
+
+    if [ -z "$target" ] || [ -z "$snap_file" ]; then
+        echo "repack-kernel: invalid argument for prepare"
+        exit 1
+    fi
+
+    target=$(realpath "$target")
+
+    mkdir -p "$target" "$target/work" "$target/backup"
+
+    # kernel snap is huge, unpacking to current dir
+    unsquashfs -d "$target/kernel" "$snap_file"
+
+    kver="$(get_kver "$target/kernel")"
+
+    # repack initrd magic, beware
+    # assumptions: initrd is compressed with LZ4, cpio block size 512, microcode
+    # at the beginning of initrd image
+    (
+        cd "$target/kernel"
+
+        # XXX: ideally we should unpack the initrd, replace snap-boostrap and
+        # repack it using ubuntu-core-initramfs --skeleton=<unpacked> this does not
+        # work and the rebuilt kernel.efi panics unable to start init, but we
+        # still need the unpacked initrd to get the right kernel modules
+        objcopy -j .initrd -O binary kernel.efi "$target/work/initrd"
+
+        # copy out the kernel image for create-efi command
+        objcopy -j .linux -O binary kernel.efi "$target/work/vmlinuz-$kver"
+
+        cp -a kernel.efi "$target/backup/"
+    )
+
+    (
+        cd "$target/work"
```

Comment:
is it needed to cd? unmkinitramfs "$target"/work/initrd "$target"/work/unpacked-initrd